seastar/src/distributed.rs

```diff
@@ -0,0 +1,537 @@
+    ///     async fn inc(&self) {
+    ///         self.0.fetch_add(1, Ordering::SeqCst);
+    ///     }
+    /// }
+    ///
+    /// impl Service for CounterService {}
+    ///
+    /// #[seastar::test]
+    /// async fn test_map_others() {
+    ///     let counter: Arc<AtomicU32> = Default::default();
+    ///     let counter_clone = counter.clone();
+    ///     let service_maker = move || CounterService(counter_clone.clone());
+    ///     let distr = Distributed::start(service_maker).await;
+    ///     
+    ///     let futs = distr.map_others(CounterService::inc);
+    ///     join_all(futs).await;
+    ///     distr.stop().await;
+    ///     
+    ///     assert_eq!(2 * (get_count() - 1), counter.load(Ordering::SeqCst));
+    /// }
+    /// ```
+    pub fn map_others<'a, Func, Ret, Fut>(&'a self, func: Func) -> Vec<impl Future<Output = Ret>>
+    where
+        Func: Fn(&'a S) -> Fut + Send + Clone + 'static,
+        Fut: Future<Output = Ret>,
+        Ret: Send + 'static,
+    {
+        let this_shard = this_shard_id();
+        self.map_selected(func, (0..get_count()).filter(move |sh| sh.ne(&this_shard)))
+    }
+
+    /// Applies a map function only to the service instance on the provided shard.
+    ///
+    /// Spiritually, a hybrid of `seastar::distributed::map` and `seastar::distributed::invoke_on`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use std::sync::atomic::{AtomicU32, Ordering};
+    /// use std::sync::Arc;
+    /// use seastar::{get_count, Distributed, Service};
+    ///
+    /// struct CounterService(Arc<AtomicU32>);
+    ///
+    /// impl CounterService {
+    ///     async fn inc(&self) {
+    ///         self.0.fetch_add(1, Ordering::SeqCst);
+    ///     }
+    /// }
+    ///
+    /// impl Service for CounterService {}
+    ///
+    /// #[seastar::test]
+    /// async fn test_map_single() {
+    ///     let counter: Arc<AtomicU32> = Default::default();
+    ///     let counter_clone = counter.clone();
+    ///     let service_maker = move || CounterService(counter_clone.clone());
+    ///     let distr = Distributed::start(service_maker).await;
+    ///     
+    ///     for shard in 0..get_count() {
+    ///         distr.map_single(shard, CounterService::inc).await;
+    ///         assert_eq!(shard + 1, counter.load(Ordering::SeqCst));
+    ///     }
+    ///     distr.stop().await;
+    /// }
+    /// ```
+    pub fn map_single<'a, Func, Ret, Fut>(
+        &'a self,
+        shard_id: u32,
+        func: Func,
+    ) -> impl Future<Output = Ret>
+    where
+        Func: Fn(&'a S) -> Fut + Send + Clone + 'static,
+        Fut: Future<Output = Ret>,
+        Ret: Send + 'static,
+    {
+        self.map_selected(func, std::iter::once(shard_id)).remove(0)
+    }
+}
+
+impl<S: Service> Deref for Distributed<S> {
+    type Target = S;
+    fn deref(&self) -> &S {
+        let local = ffi::local(self._inner.as_ref().unwrap());
+        unsafe { &*(local as *const S) }
+    }
+}
```

Comment:
This should be an explicit `local()` method instead of a `Deref` impl. `Deref` might be confusing because it is usually implemented for "smart pointer-like" types. I don't think that you can think about `distributed` in such way, mainly because you will get references to different objects depending on the shard you are performing the dereference on.

Reply:
Sure